notes-0407-composition-structure.tex

```diff
@@ -0,0 +1,303 @@
+It works similarly to homogeneous composition, but it allows the type of the cap to differ from the missing face, which explains the terminology \emph{heterogeneous} composition.
+
+There are several different variants to achieve heterogeneous composition; for instance, redtt and cubical Agda use different ones.
+Nevertheless, they both accommodate all the essential features which we are concerned with such as univalence and higher inductive types.
+
+\subsection{An Example}
+Let us recall the example of homogeneous composition given previously. 
+We may similarly apply heterogeneous composition instead of homogeneous composition except that the type $A$ can now depend on $k$. 
+
+Then, because $M$ has the type $A(0/k)$, we have
+\begin{equation*}
+\mathsf{comp}^k \ A \ [j = 0 \mapsto N_1, i = 1 \mapsto N_2, j = 1 \mapsto N_3] \ M : A(1/k),
+\end{equation*}
+where note in particular that the constructed face has the type $A(1/k)$.
+
+
+
+
+
+
+
+
+
+\section{Constraints in Contexts}
+In Agda, contexts may contain a propositional formula $\varphi$, where an atomic formula is $r = 0$ or $r = 1$ such that $r$ is an element of the De Morgan algebra given by
+\begin{equation*}
+r := 0 \mid 1 \mid i \mid r_1 \wedge r_2 \mid r_1 \vee r_2 \mid \sim \! r
+\end{equation*} 
+where $i : \mathbb{I}$.
+We then write 
+\begin{equation*}
+\Gamma \vdash M \ [\varphi \mapsto N] : A
+\end{equation*}
+to mean
+\begin{equation*}
+\Gamma, \varphi \vdash M \equiv N : A.
+\end{equation*}
+
+In this manner, we may systematically specify constraints on faces. 
+
+%There exists a functor from dimension expressions to constraints that sends $r$ to $r=1$. We can transform any constraint $\varphi$ to the form $r=1$ for some $r$.
+% NOTE(remexre): I'm not sure this follows from that functor existing; I think it needs to be strengthened to an equivalence? Or, really, just the functor in the opposite direction.
+%For example, $(i=0)\lor(i=1) = (\texttildelow i=1)\lor(i=1) = (\texttildelow i\lor i)=1$.
+
+%This lets us specify $\mathtt{transp}$ with $\oftype{r}{\mathbb{I}}$, rather than needing to be able to specify a constraint.
+
+
+
+
+\section{Kan Operators on Familiar Types}
+In this last section, let us see how homogeneous composition and coercion are implemented in standard type constructions such as unit-, empty-, pi-, sigma- and path-types.
+\subsection{Unit}
+
+\begin{multicols}{2}
+	\begin{prooftree*}
+		\hypo{\oftype{M}{\unittype}}
+		\hypo{\oftype{\varphi,\oftype{i}{\mathbb{I}}}{N}{\unittype}}
+		\infer2{\eqterm{\mathtt{hcomp}^i\,\unittype\,[\varphi\mapsto N]\,M}{M}{\unittype}}
+	\end{prooftree*}
+
+	\begin{prooftree*}
+		\hypo{\oftype{M}{\unittype}}
+		\hypo{\oftype{r}{\mathbb{I}}}
+		\infer2{\eqterm{\mathtt{transp}^i\,\unittype\,r\,M}{M}{\unittype}}
+	\end{prooftree*}
+\end{multicols}
+
+Because of the unit type's simple structure, nothing particularly interesting can happen here.
+
+\subsection{Empty}
```

Comment:
Did you mean the unit type or the empty type? BTW, the empty type should be done in the same way as other inductively defined types.